lyricsheets/effect/template_effect.py

```diff
@@ -0,0 +1,361 @@
+from collections.abc import Sequence, Mapping
+
+from ..ass.to_ass import *
+from enum import Enum
+from typing import TypeVar
+from ..models.karaoke import KObject
+
+import ast
+import re
+import operator
+
+Template = TypeVar("Template", bound="Template")
+
+inlinePattern = re.compile(r"(\$[a-zA-Z]+)")
+codePattern = re.compile(r"(![^!]+!)")
+
+
+DEFAULT_STYLE = pyass.Style.parse("BLACK,Avenir Next LT Pro,60,&H00FFFFFF,&H000000FF,&H00000000,&H00000000,-1,0,0,0,100,100,0,0,1,4,2,2,246,246,54,1")
+
+class TemplateType(Enum):
+    LINE = 1
+    PRELINE = 2
+    SYL = 3
+    CHAR = 4
+
+
+@dataclass
+class Range:
+    ranges: Sequence[tuple[int, int]]
+    unbound: bool
+
+    def __init__(self, *ranges: Sequence[tuple[int, int]]):
+        if len(ranges) == 0:
+            self.unbound = True
+            return
+        
+        self.ranges = []
+        for r in ranges:
+            if isinstance(r, tuple) and len(r) == 2:
+                start, end = r
+                if start <= end:
+                    self.ranges.append((start, end))
+                else:
+                    raise ValueError(
+                        "Range start must be less than or equal to range end"
+                    )
+            else:
+                raise ValueError("Each range must be a tuple (start, end)")
+
+    def contains(self, number: int) -> bool:
+        if self.unbound:
+            return True
+        # Check if the number falls within any of the defined ranges
+        for start, end in self.ranges:
+            if start <= number <= end:
+                return True
+        return False
```

Comment:
This looks like a reimplementation of https://docs.python.org/3/library/stdtypes.html#range but with unboundness, feels like a `list[range]` with an empty check can probably implement all this natively

Reply:
Oops, I got so used to using `range` as a generator that I forgot it could be used for... a range. Updated.